### ISSUE TITLE
501 error returned for POST requests containing upgrade header after deploying Haproxy 2.4

### ISSUE THREAD

**Issue by Akkowicz** [haproxy/haproxy#1381]
### Detailed Description of the Problem

Hi, we are using HAProxy as a loadbalancer for our internal HTTP services, after upgrading our staging installation from 2.2 to HAProxy 2.4, we've noticed that a bulk of our internal services were not able to communicate via HAProxy.

Application log:
```
feign.FeignException$NotImplemented: [501 OK] during [POST] to [http://REDACTED/internal-api/profiles/register/anonymous] [REDACTEDt#registerAnonymousProfile(RegisterAnonymousProfileRequest)]: [<html><body><h1>501 Not Implemented</h1>
.The server does not support the functionality required to fulfill the request.
</body></html>
]
```

Haproxy log:
```
staging haproxy[348179]: 127.0.0.1:33086 [26/Aug/2021:12:19:42.528] http_internal http_internal/<NOSRV> -1/-1/-1/-1/0 501 0 - - PR-- 101/101/0/0/0 0/0 "<BADREQ>"
```

After a short investigation we've found, that only the POST requests started receiving 501 errors. 
We are using Feign for building our HTTP client, it automatically includes Upgrade: h2c and Connection: HTTP2-Settings and tries to use HTTP/2 with HTTP/1.1 fallback.

### Expected Behavior

HAProxy shouldn't return 501 errors, if protocol switch is not possible it should just serve it as HTTP/1.1 fallback.

### Steps to Reproduce the Behavior

I have simple python Echo server running, I'm selecting backend based on target-service header.
The request passes correctly without --http2 option in curl.

```
curl -v --http2 --location --request POST '127.0.0.1:20000/internal-api/translations/search' --header 'Connection: HTTP2-Settings' --header 'Upgrade: h2c' --header 'target-service: bogus' --header 'Content-Type: application/json' --data-raw '{
    "test": "test"
}'
```

```
*   Trying 127.0.0.1:20000...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 20000 (#0)
> POST /internal-api/translations/search HTTP/1.1
> Host: 127.0.0.1:20000
> User-Agent: curl/7.68.0
> Accept: */*
> Connection: Upgrade, HTTP2-Settings
> Upgrade: h2c
> HTTP2-Settings: AAMAAABkAARAAAAAAAIAAAAA
> Connection: HTTP2-Settings
> Upgrade: h2c
> target-service: bogus
> Content-Type: application/json
> Content-Length: 22
> 
* upload completely sent off: 22 out of 22 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 501 Not Implemented
< Content-length: 136
< Cache-Control: no-cache
< Content-Type: text/html
< 
<html><body><h1>501 Not Implemented</h1>
.The server does not support the functionality required to fulfill the request.
</body></html>
* Connection #0 to host 127.0.0.1 left intact

```

### Do you have any idea what may have caused this?

Probably introduced in: https://github.com/haproxy/haproxy/commit/1d2d77b27f0ed5e05c4fc6b6fdc22fdb16589eef

### Do you have an idea how to solve the issue?

Maybe limit the 501 response to websocket upgrade requests only as the commit message suggests that only those are problematic (?).
Or add some option to ignore checking it.

### What is your configuration?

```haproxy
Simple HTTP loadbalancer with few backends and basic ACLs.
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.2-1ppa1~focal 2021/07/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.2.html
Running on: Linux 5.4.0-80-generic #90-Ubuntu SMP Fri Jul 9 22:49:44 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-IXRjPg/haproxy-2.4.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=1).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

### Additional Information

Ubuntu 20.04, HAProxy installed from PPA

**Comment by wtarreau** [comment]
Interesting, thanks for the report. I remember that by then we had a good reason for doing this but I agree that we could also drop the Upgrade header in this case and stay in 1.1.

I see where it is and now I remember the change. Previously it would not perform the upgrade. Now it is possible to do it except when there is a payload (too complex to pass between mux buffers). I'm not sure how it needs to be modified so I cannot provide you with a patch to try yet, I'll wait for @capflam next week. It could end up in just a few lines to just ignore the upgrade, I'm just not sure how to delete the header from there.

**Comment by chipitsine** [comment]
POST queries are used, for example, in SignalR framework. They use POST .... because it is prohibited to cache (and that is desired behaviour).


maybe, if SignalR framework has automated testing we can try to run in on top of HAProxy ?

**Comment by capflam** [comment]
Well removing the `Upgrade` header is possible for requests with a payload,  but `upgrade` token must also be removed from `Connection` header.  This part is a bit more painful. It may be a good reason to handle `upgrade` connection token as a flag only, just like `keep-alive` and `close` ones. I will investigate.

**Comment by wtarreau** [comment]
The most important part remains the removal of Upgrade, because a connection header without upgrade doesn't warrant anything since there's no protocol mentioned. Connection only enumerates a list of tokens that are hop-by-hop headers, so it's perfectly fine (even though ugly) to have upgade still listed in connection. I know that this could break on very old and unsupported haproxy versions (1.5 or so) but we don't really care. Thus it could be a first step to just drop upgrade at least to fix the current problem, and for connection we can see later for an elegant solution.

**Comment by capflam** [comment]
Here are 2 patches removing any `Upgrade` headers from requests with a payload. The first one set the header value to NULL, the second one is a HTX change to ignore such headers.

[0001-BUG-MEDIUM-mux-h1-Remove-Upgrade-header-for-requests.patch.txt](https://github.com/haproxy/haproxy/files/7136565/0001-BUG-MEDIUM-mux-h1-Remove-Upgrade-header-for-requests.patch.txt)
[0002-MINOR-htx-Skip-headers-with-no-value-when-adding-a-h.patch.txt](https://github.com/haproxy/haproxy/files/7136566/0002-MINOR-htx-Skip-headers-with-no-value-when-adding-a-h.patch.txt)

Note that for now, the `Connection` header is not updated. Thus `upgrade` token is not removed and is sent to the server.

**Comment by wtarreau** [comment]
Warning Christopher, I fear that your second patch will drop headers with an empty value. These are valid and are mandatory for `Host` when forwarding from HTTP/1.0 with unknown authority to HTTP/1.1 for example, so we must not drop headers with an empty value.

**Comment by capflam** [comment]
I know, and it should be safe because only NULL values are concerned, not empty value (len == 0 but ptr not NULL). NULL value is only possible at this stage only if it is removed after the message parsing. If an empty value is parsed, a empty string is referenced. Thus it is not a problem. But to be sure, I will check again :)

**Comment by wtarreau** [comment]
That's why I preferred to warn. Ideally if you could mention in the commit and/or comments that empty values do still work, that would save us some gray hair later :-)

**Comment by capflam** [comment]
The commit message already mentions it and there is a comment where the header is skipped. I may also update function comment to be sure to not miss the other one.

**Comment by capflam** [comment]
I pushed the fix, thus it should be fixed for the 2.5.

**Comment by capflam** [comment]
backported to 2.4 !
